### PR TITLE
1010 improve map app performance 1

### PR DESCRIPTION
**Pull request by mmnoo**
Work done to improve performance:
- fix any lint errors or warnings (will configure git hooks to lint later if I stick around and once the urgent deadline has passed)
- when fetching data with a while loop (to fetch data in chunks), only update the react state once the while loop finishes, rather than causing `n` rerenders for each chunk fetched. (the api uses pagination to control how it fetched data in chunks)
- I also got rid of a lot of data fetching that was being caused by dev server hot reloads (since we are hitting prod). This ultimately may have not been urgent, but since I brought down prod accidentally tonight, I am feeling kinda like a stickler. 
- changing how we update and store api results caused a bit of a rabbit hole with the LoadingIndicator. I moved its visibility state to live within the loading indicator component and ensure that it doesnt show when the app hot reloads 

BAsically my main task was to reduce how much I update used-all-over state in a while loop, and it kind of multiplied in unexpected ways.

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -6,11 +6,13 @@ import mermaidMuiThemeConfig from './styles/mermaidMuiThemeConfig.js'
 import MermaidDash from './components/MermaidDash/MermaidDash.jsx'
 import { Auth0Provider } from '@auth0/auth0-react'
 import { FilterProjectsProvider } from './context/FilterProjectsContext'
+import { useState } from 'react'
 
 const App = () => {
   const navigateTo = useNavigate()
   const theme = createTheme(mermaidMuiThemeConfig)
   const location = useLocation()
+  const [isApiDataDoneLoading, setIsApiDataDoneLoading] = useState(false)
```
Comment (mmnoo) [inline src/App.jsx:15]: I moved this to global app state to avoid hitting prod with hot reloading when we change the filter context file.

File: src/components/FilterPane/FilterPane.jsx [hunk 2]
```diff
@@ -102,7 +102,7 @@ const FilterPane = ({ mermaidUserData }) => {
     countriesSelectOnOpen,
     organizationsSelectOnOpen,
     getActiveProjectCount,
-  } = useFilterProjectsContext()
+  } = useContext(FilterProjectsContext)
   const [expandedSections, setExpandedSections] = useState({
```
Comment (mmnoo) [inline src/components/FilterPane/FilterPane.jsx:106]: There are a few files with this change. The linter didnt like the custom hook being defined in the context file, so I just used the more modern way to access context directly.

File: src/components/MermaidDash/MermaidDash.jsx [hunk 3]
```diff
@@ -64,30 +73,36 @@ const MermaidDash = () => {
           }
 
           const data = await response.json()
-          setProjectData((prevData) => {
```
Comment (mmnoo) [inline src/components/MermaidDash/MermaidDash.jsx:67]: we were previously setting react state that renders many many things within a while loop. Now we are just setting some state that impacts the loader and nothing else.

File: src/components/MermaidDash/MermaidDash.jsx [hunk 4]
```diff
@@ -214,7 +229,7 @@ const MermaidDash = () => {
     )
   }
 
-  const renderMap = () => (
+  const map = (
```
Comment (mmnoo) [inline src/components/MermaidDash/MermaidDash.jsx:232]: Nested react components can cause performance issues. I converted a couple to simpler JSX elements. In my next PR I will be auditing the code for more nested components and converting them to JSX if possible. (also this can cause bugs, not just performance issues)

File: src/components/MermaidDash/components/LoadingIndicator.jsx [hunk 5]
```diff
@@ -49,24 +49,25 @@ const StyledHeader = styled.header`
...
   const _calculateCurrentLoadingPercentage = useEffect(() => {
-    if (!projectData.count || !projectData.results) {
-      return
+    if (totalProjectsCount === 0 || !totalProjectsCount) {
+      return // we cant divide by zero
     }
-    setLoadingProgressValue(Math.floor((projectData.results.length / projectData.count) * 100))
-  }, [projectData])
+    setLoadingProgressValue(Math.floor((loadedProjectsCount / totalProjectsCount) * 100))
+  }, [loadedProjectsCount, totalProjectsCount])
```
Comment (mmnoo) [inline src/components/MermaidDash/components/LoadingIndicator.jsx:62]: Since we dont update the apiData to react state until its done loading now, we need something other simpler state to track progress. I didnt realize the rabbit hole I was creating until after I updated the fetching logic and noticed the progress indicator when from 0 to 100% with nothing in between.

File: src/components/MetricsPane/MetricsPane.styles.js [hunk 6]
```diff
@@ -21,7 +21,7 @@ export const StyledMetricsWrapper = styled.div`
     z-index: 5;
     background-color: transparent;
     width: 90%;
-    bottom: ${(props) => (props.$showLoadingIndicator ? '5rem;' : '0.5rem;')}
```
Comment (mmnoo) [inline src/components/MetricsPane/MetricsPane.styles.js:24]: I think these conditional styles based on the loading indicator were vestigial. I cant see them impacting anything visually anymore and removing this helped me move the state back into the loading component instead of in a parent (that is already doing a lot)

File: src/context/FilterProjectsContext.jsx [hunk 7]
```diff
@@ -641,8 +641,6 @@ export const FilterProjectsProvider = ({ children }) => {
   )
 }
 
-export const useFilterProjectsContext = () => useContext(FilterProjectsContext)
```
Comment (mmnoo) [inline src/context/FilterProjectsContext.jsx:644]: This linter didnt like this line.

File: src/components/MermaidDash/MermaidDash.jsx [hunk 8]
```diff
@@ -64,28 +74,30 @@ const MermaidDash = () => {
...
+          }
+          const resultsProjectIds = data.results.map((project) => project.project_id)
+
+          newApiData = {
+            ...newApiData,
+            count: data.count,
+            results: [...newApiData.results, ...data.results],
+          }
+          newCheckedProjects = [...newCheckedProjects, ...resultsProjectIds]
+          setLoadedProjectCount((prevCount) => prevCount + data.results.length)
```
[the remark this answers is not in the document]
Comment (mmnoo) [reply src/components/MermaidDash/MermaidDash.jsx:88]: Good question. No. :) 

So for this app we are using the pagination function of the server to load the data in chunks and display a progress loader with % loading complete displayed. We need to add up the items count returned with previous count value/sum from previous loads. 

If there is better naming to make this all more intuitive, shoot (although my scope is just performance improvements although I havent purely stuck to that scope, lol)

File: src/components/MermaidDash/MermaidDash.jsx [hunk 9]
```diff
@@ -46,10 +49,17 @@ const MermaidDash = () => {
 
   const fetchData = useCallback(
     async (token = '') => {
+      if (isApiDataDoneLoading) {
+        return
+      }
       try {
+        setLoadedProjectCount(0)
+        setTotalProjectsCount(0)
```
[the remark this answers is not in the document]
Comment (mmnoo) [reply src/components/MermaidDash/MermaidDash.jsx:57]: See related reply below. Since we are _summing_ the counts for all the calls to get chunks of data, we need to reinitialize it to 0 when we start a series of calls again, so we arent adding to numbers from previous calls.

File: src/components/MermaidDash/components/LoadingIndicator.jsx [hunk 10]
```diff
@@ -49,24 +49,25 @@ const StyledHeader = styled.header`
   `)}
 `
 
-const LoadingIndicator = ({ projectData, showLoadingIndicator, setShowLoadingIndicator }) => {
+const LoadingIndicator = ({ loadedProjectsCount, totalProjectsCount, isApiDataDoneLoading }) => {
```
[the remark this answers is not in the document]
Comment (mmnoo) [reply src/components/MermaidDash/components/LoadingIndicator.jsx:52]: Oh that is a very good call. Will you let me do this on a different branch to avoid merge conflicts? (Ive got a series of branches branched of this one and its descendants.)

**Comment by mmnoo** [comment]
[Map seems to get more laggy the more you use it](https://trello.com/c/G8Is3PyW/1010-map-seems-to-get-more-laggy-the-more-you-use-it)
